extensions/HyperVExtension/src/HyperVExtension/Services/DownloaderService.cs

```diff
@@ -12,33 +14,90 @@ namespace HyperVExtension.Services;
 /// </summary>
 public class DownloaderService : IDownloaderService
 {
+    private readonly ILogger _log = Log.ForContext("SourceContext", nameof(DownloaderService));
+
     // Use the same default buffer size as the DefaultCopyBufferSize variable in the .Nets System.IO.Stream class
     // See: https://github.com/dotnet/runtime/blob/f0117c96ace4d475af63bce80d8afa31a740b836/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L128C46-L128C52
     // For comments on why this size was chosen.
-    private const int _transferBufferSize = 81920;
+    private const int TransferBufferSize = 81920;
 
     private readonly IHttpClientFactory _httpClientFactory;
 
+    private readonly Dictionary<string, FileDownloadMonitor> _destinationFileDownloadMap = new();
+
+    private readonly object _lock = new();
+
     public DownloaderService(IHttpClientFactory httpClientFactory)
     {
         _httpClientFactory = httpClientFactory;
     }
 
     /// <inheritdoc cref="IDownloaderService.StartDownloadAsync"/>
-    public async Task StartDownloadAsync(IProgress<IOperationReport> progressProvider, Uri sourceWebUri, string destinationFile, CancellationToken cancellationToken)
-    {
-        var httpClient = _httpClientFactory.CreateClient();
-        var totalBytesToReceive = GetTotalBytesToReceive(await httpClient.GetAsync(sourceWebUri, HttpCompletionOption.ResponseHeadersRead, cancellationToken));
-        var webFileStream = await httpClient.GetStreamAsync(sourceWebUri, cancellationToken);
-        using var outputFileStream = File.OpenWrite(destinationFile);
-        outputFileStream.SetLength(totalBytesToReceive);
+    public async Task StartDownloadAsync(
+        IProgress<IOperationReport> progressProvider,
+        Uri sourceWebUri,
+        string destinationFilePath,
+        CancellationToken cancellationToken)
+    {
+        var downloadMonitor = new FileDownloadMonitor(progressProvider);
 
-        var downloadProgress = new Progress<ByteTransferProgress>(progressObj =>
+        try
         {
-            progressProvider.Report(new DownloadOperationReport(progressObj));
-        });
+            lock (_lock)
```

Comment:
This lock is taken for the duration of the whole async task. Lock/unlock should happen on the same thread, but when execution returns to this code after one of the awaits it may not return to the same thread.

["You can't use the await expression in the body of a lock statement."](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock#:~:text=You%20can%27t%20use%20the%20await%20expression%20in%20the%20body%20of%20a%20lock%20statement.)

Reply:
I'm not sure I follow here. I'm not actually using an await statement in the lock. my intent isn't to lock the whole method, only the start and completion of the method to access the `_destinationFileDownloadMap`. In both places I don't use await within the lock, but please let me know if I'm misunderstanding something, thanks.